packages/core/src/sync-store/postgres/store.ts

```diff
@@ -539,23 +551,178 @@ export class PostgresSyncStore implements SyncStore {
+      });
+    });
+  };
+
+  getBlockFilterIntervals = async ({
+    chainId,
+    blockFilter,
+  }: {
+    chainId: number;
+    blockFilter: BlockFilterCriteria;
+  }) => {
+    return this.db.wrap({ method: "getBlockFilterIntervals" }, async () => {
+      const fragment = {
+        id: `${chainId}_${blockFilter.frequency}_${blockFilter.offset}`,
+        chainId,
+        frequency: blockFilter.frequency,
+        offset: blockFilter.offset,
+      };
+
+      // First, attempt to merge overlapping and adjacent intervals.
+      await this.db.transaction().execute(async (tx) => {
+        const { id: blockFilterId } = await tx
+          .insertInto("blockFilters")
+          .values(fragment)
+          .onConflict((oc) => oc.column("id").doUpdateSet(fragment))
+          .returningAll()
+          .executeTakeFirstOrThrow();
+
+        const existingIntervalRows = await tx
+          .deleteFrom("blockFilterIntervals")
+          .where("blockFilterId", "=", blockFilterId)
+          .returningAll()
+          .execute();
+
+        const mergedIntervals = intervalUnion(
+          existingIntervalRows.map((i) => [
+            Number(i.startBlock),
+            Number(i.endBlock),
+          ]),
+        );
+
+        const mergedIntervalRows = mergedIntervals.map(
+          ([startBlock, endBlock]) => ({
+            blockFilterId,
+            startBlock: BigInt(startBlock),
+            endBlock: BigInt(endBlock),
+          }),
+        );
+
+        if (mergedIntervalRows.length > 0) {
+          await tx
+            .insertInto("blockFilterIntervals")
+            .values(mergedIntervalRows)
+            .execute();
+        }
+
+        return mergedIntervals;
+      });
+
+      const intervals = await this.db
+        .selectFrom("blockFilterIntervals")
+        .innerJoin("blockFilters", "blockFilterId", "blockFilters.id")
+        .select([
+          "blockFilterId",
+          "blockFilterIntervals.startBlock",
+          "blockFilterIntervals.endBlock",
+        ])
+        .where("chainId", "=", chainId)
+        .where("blockFilters.frequency", "=", fragment.frequency)
+        .where("blockFilters.offset", "=", fragment.offset)
+        .execute();
+
+      const intervalsByFragmentId = intervals.reduce(
+        (acc, cur) => {
+          const { blockFilterId, startBlock, endBlock } = cur;
+          (acc[blockFilterId] ||= []).push([
+            Number(startBlock),
+            Number(endBlock),
+          ]);
+          return acc;
+        },
+        {} as Record<string, [number, number][]>,
+      );
+
+      return intervalUnion(intervalsByFragmentId[fragment.id] ?? []);
+    });
+  };
+
+  getBlock = async ({
+    chainId,
+    blockNumber,
+  }: {
+    chainId: number;
+    blockNumber: number;
+  }): Promise<boolean> => {
+    const hasBlock = await this.db
+      .selectFrom("blocks")
+      .select("hash")
+      .where("number", "=", BigInt(blockNumber))
+      .where("chainId", "=", chainId)
```

Comment:
Great question - seems like this would be prone to loading an incorrect block if:

1) a block is indexed in realtime
2) app shuts down
3) app starts back up, but the re-orged block is now considered finalized

It might be so over for this feature until we can figure out a way to tell if a block in the database is finalized. Maybe a network_metadata table but that would probably be in another release